File: app/auth/session.py
from __future__ import annotations

import base64
import hashlib
import hmac
import os
import time
from typing import Any

import jwt


SESSION_COOKIE = "costgate_session"
SESSION_TTL_SECONDS = int(os.getenv("SESSION_TTL_SECONDS", str(60 * 60 * 24 * 7)))


def _secret() -> str:
    secret = os.getenv("AUTH_JWT_SECRET")
    if not secret or len(secret) < 32:
        raise RuntimeError("AUTH_JWT_SECRET must be configured with at least 32 characters.")
    return secret


def create_session_token(*, user_id: str) -> str:
    now = int(time.time())
    payload = {
        "sub": user_id,
        "iat": now,
        "exp": now + SESSION_TTL_SECONDS,
    }
    return jwt.encode(payload, _secret(), algorithm="HS256")


def verify_session_token(token: str) -> dict[str, Any] | None:
    try:
        return jwt.decode(token, _secret(), algorithms=["HS256"])
    except jwt.PyJWTError:
        return None


def parse_cookie_header(cookie_header: str | None, name: str = SESSION_COOKIE) -> str | None:
    if not cookie_header:
        return None
    for pair in cookie_header.split(";"):
        key, _, value = pair.strip().partition("=")
        if key == name and value:
            return value
    return None


def get_user_id_from_event(event: dict[str, Any]) -> str | None:
    headers = {str(k).lower(): str(v) for k, v in (event.get("headers") or {}).items()}
    authorization = headers.get("authorization", "")
    token = authorization.removeprefix("Bearer ").strip() if authorization else ""
    if not token:
        token = parse_cookie_header(headers.get("cookie")) or ""
    if not token:
        return None
    claims = verify_session_token(token)
    subject = claims.get("sub") if claims else None
    return str(subject) if subject else None


def hash_password(password: str) -> str:
    if len(password) < 10:
        raise ValueError("Password must be at least 10 characters long.")
    salt = os.urandom(16)
    derived = hashlib.scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=2**14,
        r=8,
        p=1,
        dklen=64,
    )
    return "scrypt$16384$8$1$" + _b64(salt) + "$" + _b64(derived)


def verify_password(password: str, encoded: str) -> bool:
    try:
        scheme, n, r, p, salt_b64, hash_b64 = encoded.split("$", 5)
        if scheme != "scrypt":
            return False
        derived = hashlib.scrypt(
            password.encode("utf-8"),
            salt=_unb64(salt_b64),
            n=int(n),
            r=int(r),
            p=int(p),
            dklen=len(_unb64(hash_b64)),
        )
        return hmac.compare_digest(derived, _unb64(hash_b64))
    except (ValueError, TypeError):
        return False


def _b64(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _unb64(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))

File: frontend/lib/auth/session.ts
import { jwtVerify, SignJWT } from "jose";
import { cookies } from "next/headers";

export const SESSION_COOKIE = "costgate_session";

const secretValue = process.env.AUTH_JWT_SECRET;

if (!secretValue) {
    throw new Error("AUTH_JWT_SECRET is not configured");
}

if (secretValue.length < 32) {
    throw new Error(
        "AUTH_JWT_SECRET must be at least 32 characters"
    );
}

const secret = new TextEncoder().encode(secretValue);

const ISSUER = "costgate";
const AUDIENCE = "costgate-web";

export type Session = {
    userId: string;
    email: string;
    name: string;
};

export async function createSessionToken(
    session: Session
): Promise<string> {
    return new SignJWT({
        sub: session.userId,
        email: session.email,
        name: session.name,
    })
        .setProtectedHeader({
            alg: "HS256",
        })
        .setIssuedAt()
        .setExpirationTime("7d")
        .setIssuer(ISSUER)
        .setAudience(AUDIENCE)
        .sign(secret);
}

export async function verifySessionToken(
    token: string
): Promise<Session | null> {
    try {
        const { payload } = await jwtVerify(token, secret, {
            algorithms: ["HS256"],
        });

        if (typeof payload.sub !== "string" || !payload.sub) {
            return null;
        }

        return {
            userId: payload.sub,
            email: typeof payload.email === "string" ? payload.email : "",
            name: typeof payload.name === "string" ? payload.name : "",
        };
    } catch {
        return null;
    }
}

export async function getSession(): Promise<Session | null> {
    const cookieStore = await cookies();

    const token = cookieStore.get(
        SESSION_COOKIE
    )?.value;

    if (!token) {
        return null;
    }

    return verifySessionToken(token);
}

export async function getSessionToken() {
    const cookieStore =
        await cookies();

    return (
        cookieStore.get(
            SESSION_COOKIE,
        )?.value ?? null
    );
}